db/commitlog/commitlog.cc

```diff
@@ -3135,6 +3555,100 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, commit_load_reader_f
                     }
                 }
 
+                co_return;
+            } else if (size == segment::fragmented_entry_size_magic) {
+                auto actual_size = checksum;
+                auto end = next_pos(actual_size - entry_header_size);
+
+                clogger.debug("read_entry (fragmented) size = {}", actual_size);
+
+                assert(end <= next);
+
+                assert(end <= next);
+                // really small read...
+                buf = co_await read_data(segment::fragmented_entry_overhead_size);
+                in = buf.get_istream();
+
+                auto id = read<uint32_t>(in);
+                auto off = read<uint32_t>(in);
+                auto rem = read<uint32_t>(in);
+                checksum = read<uint32_t>(in);
+
+                crc.process(actual_size);
+                crc.process(id);
+                crc.process(off);
+                crc.process(rem);
+
+                if (crc.checksum() != checksum) {
+                    auto slack = next - pos;
+                    if (size != 0) {
+                        clogger.debug("Fractured segment entry at {} has broken header. Skipping to next chunk ({} bytes)", rp, slack);
+                        corrupt_size += slack;
+                    }
+                    co_await skip_to_chunk(next);
```

Comment:
How can we ignore commits like this?

Reply:
We generally skip and later report corrupted data, but continue replaying what we can gobble together. Since we cannot safely determine anything here, we unfortunately can't drop any partial data or anything, but we can at least play back whatever else might be readable.

Reply:
We need to only allow the last commit to be broken. Raft requires that durable commits be really durable, not maybe durable. The last commit can be broken if we crashed while writing it.

If this is existing practice, we can retain it and fix it up later.

Reply:
It is. Nothing is changed in the corruption management regard. 
But what do you mean "we need only allow..."? If raft requires truly durable commits (i.e. full consistency), then we cannot really allow any data to be broken, and in the case of multi-entry commits, we need to change the handling to use buffering and collect _all_ entries before trying to apply any one of them Today, we only guarantee multi-entries get _written_ "atomically". Not that they are _read_ atomically.

Reply:
I.e. data corruption on disk is allowed to partially corrupt a multi-entry, i.e. a potential raft set.

Reply:
I believe Raft will actually work without fully atomic commits, as long as we report a partial apply as failed. However I wouldn't bet on it. And if Raft batches are committed as multiple mutations to the log, then a partial apply won't work.

In any case I'm not confident enough to say it will work for sure. Better be atomic.